File: python/1031. Maximum Sum of Two Non-Overlapping Subarrays.py
class Solution(object):
    def maxSumTwoNoOverlap(self, A, L, M):
        """
        :type A: List[int]
        :type L: int
        :type M: int
        :rtype: int
        """
        n = len(A)
        max_L = 0
        max_LM = 0
        
        for i in range(n-L-M+1):
            max_L = max(max_L, sum(A[i:i+L]))
            max_LM = max(max_LM, max_L + sum(A[i+L:i+L+M]))

        max_M = 0
        max_ML = 0
        for i in range(n-L-M+1):
            max_M = max(max_M, sum(A[i:i+M]))
            max_ML = max(max_ML, max_M + sum(A[i+M:i+L+M]))            
        
        return max(max_LM, max_ML)

File: go/src/leetcode/1031/1031. Maximum Sum of Two Non-Overlapping Subarrays.go
package main

/*
*
https://leetcode.cn/problems/maximum-sum-of-two-non-overlapping-subarrays/
*/
func maxSumTwoNoOverlap(nums []int, L1 int, L2 int) (tot int) {
	n := len(nums)
	sums := make([]int, n+1)
	for i := range nums {
		sums[i+1] = sums[i] + nums[i]
	}
	s := func(r, l int) int {
		return sums[min(n, r)] - sums[max(0, l)]
	}

	var l, lMax, r, rMax int
	for i := 0; i < n-L1-L2+1; i++ {
		l = max(l, s(i+L1, i))
		lMax = max(lMax, l+s(i+L1+L2, i+L1))
	}

	for i := 0; i < n-L1-L2+1; i++ {
		r = max(r, s(i+L2, i))
		rMax = max(rMax, r+s(i+L2+L1, i+L2))
	}
	return max(lMax, rMax)
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	maxSumTwoNoOverlap([]int{0, 6, 5, 2, 2, 5, 1, 9, 4}, 1, 2)
}
